### ISSUE TITLE
Pool flooding WAIT bug

### ISSUE THREAD

**Issue by luke-jr** [luke-jr/bfgminer#85]
During a ModMiner bitstream upload, Icarus stopped hashing into WAIT state and getwork count shot up fast.

New work is being rejected immediately (inside get_work)... "Work stale due to block mismatch"

This was on Eligius.

Backtrace:

```
Thread 28 (Thread 0xf726cb70 (LWP 29193)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x0805f36d in tq_pop (tq=0x8114e00, abstime=0x0) at util.c:640
#3  0x08055d6c in workio_thread (userdata=0x8117cf0) at miner.c:3796
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 27 (Thread 0xf6a6bb70 (LWP 29194)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x0805f36d in tq_pop (tq=0x8122f70, abstime=0x0) at util.c:640
#3  0x080538dd in stage_thread (userdata=0x8117d38) at miner.c:3075
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 26 (Thread 0xf626ab70 (LWP 29197)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf755fa45 in poll () from /lib/libc.so.6
#2  0xf76d711a in Curl_socket_check () from /usr/lib/libcurl.so.4
#3  0xf76c7961 in Transfer () from /usr/lib/libcurl.so.4
#4  0xf76c87dd in Curl_do_perform () from /usr/lib/libcurl.so.4
#5  0xf76c8a6d in Curl_perform () from /usr/lib/libcurl.so.4
#6  0xf76c92b5 in curl_easy_perform () from /usr/lib/libcurl.so.4
#7  0x0805eace in json_rpc_call (curl=0x8130658, url=0x8124f18 "http://mining.eligius.st:8337/LP", 
    userpass=0x8117ba0 "16kNKa7WUg8QAPFy8dJRv7USSu2fAG2pkW:x", 
    rpc_req=0x80e1cb0 "{\"method\": \"getwork\", \"params\": [], \"id\":0}\r\n", probe=false, longpoll=true, 
    rolltime=0xf626a360, pool=0x81068d0, share=false) at util.c:369
#8  0x0805887c in longpoll_thread (userdata=0x81068d0) at miner.c:4795
#9  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#10 0xf756a3ee in clone () from /lib/libc.so.6

Thread 25 (Thread 0xf50feb70 (LWP 29199)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a8b4 in pthread_cond_timedwait () from /lib/libpthread.so.0
#2  0x08056d5d in hash_pop (abstime=0xf50fe298) at miner.c:4132
#3  0x080574bc in get_work (work=0x81255d8, requested=false, thr=0x8117bd0, thr_id=0) at miner.c:4309
#4  0x08057c7c in miner_thread (userdata=0x8117bd0) at miner.c:4508
#5  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#6  0xf756a3ee in clone () from /lib/libc.so.6

Thread 24 (Thread 0xf30fab70 (LWP 29203)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a8b4 in pthread_cond_timedwait () from /lib/libpthread.so.0
#2  0x08056d5d in hash_pop (abstime=0xf30fa298) at miner.c:4132
#3  0x080574bc in get_work (work=0x81184f0, requested=false, thr=0x8117c18, thr_id=1) at miner.c:4309
#4  0x08057c7c in miner_thread (userdata=0x8117c18) at miner.c:4508
#5  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#6  0xf756a3ee in clone () from /lib/libc.so.6

Thread 23 (Thread 0xf28f9b70 (LWP 29204)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a8b4 in pthread_cond_timedwait () from /lib/libpthread.so.0
#2  0x08056d5d in hash_pop (abstime=0xf28f9298) at miner.c:4132
#3  0x080574bc in get_work (work=0x8118288, requested=false, thr=0x8117c60, thr_id=2) at miner.c:4309
#4  0x08057c7c in miner_thread (userdata=0x8117c60) at miner.c:4508
#5  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#6  0xf756a3ee in clone () from /lib/libc.so.6

Thread 22 (Thread 0xf20f8b70 (LWP 29205)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a8b4 in pthread_cond_timedwait () from /lib/libpthread.so.0
#2  0x08056d5d in hash_pop (abstime=0xf20f8298) at miner.c:4132
#3  0x080574bc in get_work (work=0x8119558, requested=false, thr=0x8117ca8, thr_id=3) at miner.c:4309
#4  0x08057c7c in miner_thread (userdata=0x8117ca8) at miner.c:4508
#5  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#6  0xf756a3ee in clone () from /lib/libc.so.6

Thread 21 (Thread 0xf18f7b70 (LWP 29207)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf7531206 in nanosleep () from /lib/libc.so.6
#2  0xf7531024 in sleep () from /lib/libc.so.6
#3  0x08058da1 in watchpool_thread (userdata=0x0) at miner.c:4942
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 20 (Thread 0xf10f6b70 (LWP 29208)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf7531206 in nanosleep () from /lib/libc.so.6
#2  0xf7531024 in sleep () from /lib/libc.so.6
#3  0x08058e95 in watchdog_thread (userdata=0x0) at miner.c:4988
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 19 (Thread 0xf08f5b70 (LWP 29209)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x0805f36d in tq_pop (tq=0x8119318, abstime=0x0) at util.c:640
#3  0x0806f0bd in reinit_gpu (userdata=0x8117e10) at driver-opencl.c:1363
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 18 (Thread 0xf00f4b70 (LWP 29210)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770d7c8 in accept () from /lib/libpthread.so.0
#2  0x0806a783 in api (api_thr_id=9) at api.c:3184
#3  0x08055e1d in api_thread (userdata=0x8117e58) at miner.c:3829
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 17 (Thread 0xef8f3b70 (LWP 29211)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770d5bb in read () from /lib/libpthread.so.0
#2  0xf7609bb0 in fifo_push () from /lib/libncurses.so.5
#3  0xf760a349 in kgetch () from /lib/libncurses.so.5
#4  0xf760a01b in _nc_wgetch () from /lib/libncurses.so.5
#5  0xf760a27e in wgetch () from /lib/libncurses.so.5
#6  0x08055b7a in input_thread (userdata=0x8117ea0) at miner.c:3734
#7  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#8  0xf756a3ee in clone () from /lib/libc.so.6

Thread 16 (Thread 0xf48fdb70 (LWP 29213)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf755fa45 in poll () from /lib/libc.so.6
#2  0xf76d711a in Curl_socket_check () from /usr/lib/libcurl.so.4
#3  0xf76c7961 in Transfer () from /usr/lib/libcurl.so.4
#4  0xf76c87dd in Curl_do_perform () from /usr/lib/libcurl.so.4
#5  0xf76c8a6d in Curl_perform () from /usr/lib/libcurl.so.4
#6  0xf76c92b5 in curl_easy_perform () from /usr/lib/libcurl.so.4
#7  0x0805eace in json_rpc_call (curl=0x8141408, url=0x8126898 "http://mining.eligius.st:8337/LP", 
    userpass=0x8122e20 "16kNKa7WUg8QAPFy8dJRv7USSu2fAG2pkW:x", 
    rpc_req=0x80e1cb0 "{\"method\": \"getwork\", \"params\": [], \"id\":0}\r\n", probe=false, longpoll=true, 
    rolltime=0xf48fd360, pool=0x8107920, share=false) at util.c:369
#8  0x0805887c in longpoll_thread (userdata=0x8107920) at miner.c:4795
#9  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#10 0xf756a3ee in clone () from /lib/libc.so.6

Thread 15 (Thread 0xf40fcb70 (LWP 29215)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x080586c3 in wait_lpcurrent (pool=0x8107a38) at miner.c:4741
#3  0x0805879b in longpoll_thread (userdata=0x8107a38) at miner.c:4776
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 14 (Thread 0xef0f2b70 (LWP 29217)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x080586c3 in wait_lpcurrent (pool=0x8107b50) at miner.c:4741
#3  0x0805879b in longpoll_thread (userdata=0x8107b50) at miner.c:4776
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 13 (Thread 0xf58ffb70 (LWP 29219)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x080586c3 in wait_lpcurrent (pool=0x8107c68) at miner.c:4741
#3  0x0805879b in longpoll_thread (userdata=0x8107c68) at miner.c:4776
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 12 (Thread 0xf38fbb70 (LWP 29221)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x080586c3 in wait_lpcurrent (pool=0x8107d80) at miner.c:4741
#3  0x0805879b in longpoll_thread (userdata=0x8107d80) at miner.c:4776
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 11 (Thread 0xee8f1b70 (LWP 29223)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x080586c3 in wait_lpcurrent (pool=0x8107e98) at miner.c:4741
#3  0x0805879b in longpoll_thread (userdata=0x8107e98) at miner.c:4776
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 10 (Thread 0xee0f0b70 (LWP 29225)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x080586c3 in wait_lpcurrent (pool=0x8107fb0) at miner.c:4741
#3  0x0805879b in longpoll_thread (userdata=0x8107fb0) at miner.c:4776
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 9 (Thread 0xed8efb70 (LWP 29227)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x080586c3 in wait_lpcurrent (pool=0x81080c8) at miner.c:4741
#3  0x0805879b in longpoll_thread (userdata=0x81080c8) at miner.c:4776
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 8 (Thread 0xec8edb70 (LWP 31666)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x080586c3 in wait_lpcurrent (pool=0x811e138) at miner.c:4741
#3  0x0805879b in longpoll_thread (userdata=0x811e138) at miner.c:4776
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 7 (Thread 0xed0eeb70 (LWP 3574)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf755fa45 in poll () from /lib/libc.so.6
#2  0xf76d711a in Curl_socket_check () from /usr/lib/libcurl.so.4
#3  0xf76c7961 in Transfer () from /usr/lib/libcurl.so.4
#4  0xf76c87dd in Curl_do_perform () from /usr/lib/libcurl.so.4
#5  0xf76c8a6d in Curl_perform () from /usr/lib/libcurl.so.4
#6  0xf76c92b5 in curl_easy_perform () from /usr/lib/libcurl.so.4
#7  0x0805eace in json_rpc_call (curl=0x8138d30, url=0x8106bc8 "http://mining.eligius.st:8337", 
    userpass=0x8117ba0 "16kNKa7WUg8QAPFy8dJRv7USSu2fAG2pkW:x", 
    rpc_req=0x80e1cb0 "{\"method\": \"getwork\", \"params\": [], \"id\":0}\r\n", probe=false, longpoll=false, 
    rolltime=0x8125a6c, pool=0x81068d0, share=false) at util.c:369
#8  0x0804fdd1 in get_upstream_work (work=0x8125840, curl=0x8138d30) at miner.c:2136
#9  0x08050845 in get_work_thread (userdata=0xf590a8f8) at miner.c:2459
#10 0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#11 0xf756a3ee in clone () from /lib/libc.so.6

Thread 6 (Thread 0xe78e3b70 (LWP 3575)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf755fa45 in poll () from /lib/libc.so.6
#2  0xf76d711a in Curl_socket_check () from /usr/lib/libcurl.so.4
#3  0xf76c7961 in Transfer () from /usr/lib/libcurl.so.4
#4  0xf76c87dd in Curl_do_perform () from /usr/lib/libcurl.so.4
#5  0xf76c8a6d in Curl_perform () from /usr/lib/libcurl.so.4
#6  0xf76c92b5 in curl_easy_perform () from /usr/lib/libcurl.so.4
#7  0x0805eace in json_rpc_call (curl=0x8127f80, url=0x8106bc8 "http://mining.eligius.st:8337", 
    userpass=0x8117ba0 "16kNKa7WUg8QAPFy8dJRv7USSu2fAG2pkW:x", 
    rpc_req=0x80e1cb0 "{\"method\": \"getwork\", \"params\": [], \"id\":0}\r\n", probe=false, longpoll=false, 
    rolltime=0x811dfcc, pool=0x81068d0, share=false) at util.c:369
#8  0x0804fdd1 in get_upstream_work (work=0x811dda0, curl=0x8127f80) at miner.c:2136
#9  0x08050845 in get_work_thread (userdata=0xf590a980) at miner.c:2459
#10 0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#11 0xf756a3ee in clone () from /lib/libc.so.6

Thread 5 (Thread 0xe90e6b70 (LWP 3576)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf755fa45 in poll () from /lib/libc.so.6
#2  0xf76d711a in Curl_socket_check () from /usr/lib/libcurl.so.4
#3  0xf76c7961 in Transfer () from /usr/lib/libcurl.so.4
#4  0xf76c87dd in Curl_do_perform () from /usr/lib/libcurl.so.4
#5  0xf76c8a6d in Curl_perform () from /usr/lib/libcurl.so.4
#6  0xf76c92b5 in curl_easy_perform () from /usr/lib/libcurl.so.4
#7  0x0805eace in json_rpc_call (curl=0x81b57b8, url=0x8106bc8 "http://mining.eligius.st:8337", 
    userpass=0x8117ba0 "16kNKa7WUg8QAPFy8dJRv7USSu2fAG2pkW:x", 
    rpc_req=0x80e1cb0 "{\"method\": \"getwork\", \"params\": [], \"id\":0}\r\n", probe=false, longpoll=false, 
    rolltime=0x8125eb4, pool=0x81068d0, share=false) at util.c:369
#8  0x0804fdd1 in get_upstream_work (work=0x8125c88, curl=0x81b57b8) at miner.c:2136
#9  0x08050845 in get_work_thread (userdata=0xe9f00560) at miner.c:2459
#10 0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#11 0xf756a3ee in clone () from /lib/libc.so.6

Thread 4 (Thread 0xe80e4b70 (LWP 3577)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf755fa45 in poll () from /lib/libc.so.6
#2  0xf76d711a in Curl_socket_check () from /usr/lib/libcurl.so.4
#3  0xf76c7961 in Transfer () from /usr/lib/libcurl.so.4
#4  0xf76c87dd in Curl_do_perform () from /usr/lib/libcurl.so.4
#5  0xf76c8a6d in Curl_perform () from /usr/lib/libcurl.so.4
#6  0xf76c92b5 in curl_easy_perform () from /usr/lib/libcurl.so.4
#7  0x0805eace in json_rpc_call (curl=0x81bde90, url=0x8106bc8 "http://mining.eligius.st:8337", 
    userpass=0x8117ba0 "16kNKa7WUg8QAPFy8dJRv7USSu2fAG2pkW:x", 
    rpc_req=0x80e1cb0 "{\"method\": \"getwork\", \"params\": [], \"id\":0}\r\n", probe=false, longpoll=false, 
    rolltime=0x812711c, pool=0x81068d0, share=false) at util.c:369
#8  0x0804fdd1 in get_upstream_work (work=0x8126ef0, curl=0x81bde90) at miner.c:2136
#9  0x08050845 in get_work_thread (userdata=0xe9f01170) at miner.c:2459
#10 0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#11 0xf756a3ee in clone () from /lib/libc.so.6

Thread 3 (Thread 0xea8e9b70 (LWP 3578)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf755fa45 in poll () from /lib/libc.so.6
#2  0xf76d711a in Curl_socket_check () from /usr/lib/libcurl.so.4
#3  0xf76c7961 in Transfer () from /usr/lib/libcurl.so.4
#4  0xf76c87dd in Curl_do_perform () from /usr/lib/libcurl.so.4
#5  0xf76c8a6d in Curl_perform () from /usr/lib/libcurl.so.4
#6  0xf76c92b5 in curl_easy_perform () from /usr/lib/libcurl.so.4
#7  0x0805eace in json_rpc_call (curl=0x81ad0e0, url=0x8106bc8 "http://mining.eligius.st:8337", 
    userpass=0x8117ba0 "16kNKa7WUg8QAPFy8dJRv7USSu2fAG2pkW:x", 
    rpc_req=0x80e1cb0 "{\"method\": \"getwork\", \"params\": [], \"id\":0}\r\n", probe=false, longpoll=false, 
    rolltime=0x811bab4, pool=0x81068d0, share=false) at util.c:369
#8  0x0804fdd1 in get_upstream_work (work=0x811b888, curl=0x81ad0e0) at miner.c:2136
#9  0x08050845 in get_work_thread (userdata=0xf5901d18) at miner.c:2459
#10 0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#11 0xf756a3ee in clone () from /lib/libc.so.6

Thread 2 (Thread 0xe9effb70 (LWP 3579)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf770a4fc in pthread_cond_wait () from /lib/libpthread.so.0
#2  0x0805054a in pop_curl_entry (pool=0x81068d0) at miner.c:2363
#3  0x080507b0 in get_work_thread (userdata=0xf59012a8) at miner.c:2456
#4  0xf7705ed3 in start_thread () from /lib/libpthread.so.0
#5  0xf756a3ee in clone () from /lib/libc.so.6

Thread 1 (Thread 0xf726d6c0 (LWP 29192)):
#0  0xffffe430 in __kernel_vsyscall ()
#1  0xf7706fe6 in pthread_join () from /lib/libpthread.so.0
#2  0x0805d6be in main (argc=1, argv=0xffd739b4) at miner.c:6106
```

**Comment by luke-jr** [comment]
Hopefully fixed by d19879e

**Comment by cavalierro** [comment]
Hi, I still get this sometimes with bfgminer-2.10.5 (openwrt) and p2pool (terracoin ). 
In my case, bfgminer gets into this if freshly started and immediately gets a stratum work restart.
(edit):
Messages: Debug  "Work stale due to block mismatch (f59904bd != 0 ? f59904bd : ff606682)"
And normal output only "Stratum from pool 0 requested work restart"

This also happens sometimes when switching to a new stratum pool.

**Comment by cavalierro** [comment]
More precisely: Debug  "Work stale due to block mismatch (f59904bd != 0 ? f59904bd : ff606682)"
And normal output only "Stratum from pool 0 requested work restart"

This also happens sometimes when switching to a new stratum pool.

**Comment by peeweed44** [comment]
This has become an increasingly common occurrence for me.   To make matters worse any restarts or pool switches after it hangs in wait restarts it in the same condition.   Let one worker go for over 5 hours the other day and never recovered.

**Comment by luke-jr** [comment]
This bug is fixed. Your issue has nothing to do with it as far as I can see.
Open a new issue with full details of your problem.

**Comment by cavalierro** [comment]
You cand leave it closed, but I'll give here the solution, since this is the thread I found when faced with the problem. 

I'm not sure if it's a bug, I can open one no problem if necessary.

Basically it happens because I mixed TRC + BTC pools in the pool options, while having strategy Failover but "Failover only" = not enabled (the default). In this mode the miner sends shares to different coin pools, and there is the conflict.

I can solve it by either disabling the other coin's pools or by using --failover-only.
